src/openslide-vendor-philips-isyntax.c

```diff
@@ -0,0 +1,397 @@
+        for (int level_idx = 0; level_idx < wsi->level_count; ++level_idx) {
+            memory_count += wsi->levels[level_idx].tile_count * sizeof(isyntax_tile_t);
+        }
+        memory_count += wsi->codeblock_count * sizeof(isyntax_codeblock_t);
+        memory_count += wsi->data_chunk_count * sizeof(isyntax_data_chunk_t);
+        printf("philips_isyntax_open is_global_cache=%d cache_size=%d sizeof_structs=%'lld\n", (int)is_global_cache, cache_size, memory_count);
+    } */
+    if (is_global_cache) {
+        g_mutex_lock(&static_open_mutex);
+        if (philips_isyntax_global_cache_ptr == NULL) {
+            // Note: this requires that all opened files have the same block size. If that is not true, we
+            // will need to have allocator per size. Alternatively, implement allocator freeing after
+            // all tiles have been freed, and track isyntax_t per tile so we can access allocator.
+            philips_isyntax_global_cache_ptr = malloc(sizeof(*philips_isyntax_global_cache_ptr));
+            ASSERT_OK(libisyntax_cache_create("global_cache_list", cache_size, &philips_isyntax_global_cache_ptr->cache));
+        }
+        data->cache = philips_isyntax_global_cache_ptr;
+        g_mutex_unlock(&static_open_mutex);
+    } else {
+        ASSERT_OK(libisyntax_cache_create("cache_list", cache_size, &data->cache->cache));
+    }
+    // Link the cache (local or global) to the isyntax file.
+    libisyntax_cache_inject(data->cache->cache, data->isyntax);
+
+    LOG_VAR("%d", libisyntax_get_is_mpp_known(data->isyntax));
+    if (libisyntax_get_is_mpp_known(data->isyntax)) {
+        double mpp_x = libisyntax_get_mpp_x(data->isyntax);
+        double mpp_y = libisyntax_get_mpp_y(data->isyntax);
+        LOG_VAR("%f", mpp_x);
+        LOG_VAR("%f", mpp_y);
+        add_float_property(osr, OPENSLIDE_PROPERTY_NAME_MPP_X, mpp_x);
+        add_float_property(osr, OPENSLIDE_PROPERTY_NAME_MPP_Y, mpp_y);
+        const float float_equals_tolerance = 1e-5;
+        if (fabs(mpp_x - mpp_y) < float_equals_tolerance) {
+            // Compute objective power from microns-per-pixel, see e.g. table in "Scan Performance" here:
+            // https://www.microscopesinternational.com/blog/20170928-whichobjective.aspx
+            float objective_power = 10.0f / mpp_x;
+            LOG_VAR("%f", objective_power);
+            add_float_property(osr, OPENSLIDE_PROPERTY_NAME_OBJECTIVE_POWER, objective_power);
+        }
+    }
+
+    // Find wsi image. Extracting other images not supported. Assuming only one wsi.
+    int wsi_image_idx = libisyntax_get_wsi_image_index(data->isyntax);
+    LOG_VAR("%d", wsi_image_idx);
+    const isyntax_image_t* wsi_image = libisyntax_get_image(data->isyntax, wsi_image_idx);
+
+    // Store openslide information about each level.
+    osr->level_count = libisyntax_image_get_level_count(wsi_image);
+    osr->levels = malloc(sizeof(philips_isyntax_level*) * osr->level_count);
```

Comment:
It'd improve legibility for this driver to follow the usual pattern of setting up its data structures and then only populating `osr` at the end.  See the other drivers for examples.

Most drivers use a `GPtrArray` to build up the level array because they don't know the level count in advance, and most drivers have to contend with I/O errors when doing so.  The first is redundant here; the second is currently redundant but may not stay that way.  I'd be inclined to recommend using `GPtrArray` for building up levels anyway, so you can just make a `g_autoptr(GPtrArray)` and not worry about cleanup in any error cases.

Reply:
Done. I agree with the `GPtrArray` generic usage, converted to that. I allocate the array `g_ptr_array_sized_new(level_count)` because I know the size.